modules/python-modules/README.md

```diff
@@ -0,0 +1,648 @@
+
+```python
+
+class MyFetcher(LogFetcher):
+
+    ...
+
+    def init(self, options):
+        # pass self.message_acked method as ACK callback
+        self.ack_tracker = ConsecutiveAckTracker(ack_callback=self.message_acked)
+        return True
+
+    def message_acked(self, acked_message_bookmark):
+        pass
+
+    def fetch(self):
+        ...
+        msg.set_bookmark("whatever-bookmark-value-that-denotes-position")
+    ...
+
+```
+
+In the case above, we were using ConsecutiveAckTracker, e.g.  we would only
+get acknowledgements in the order messages were generated.  The argument of
+the `message_acked` callback would be the "bookmark" value that we set using
+set_bookmark().
+
+Using InstantAckTracker is very similar, just replace
+`ConsecutiveAckTracker` with `InstantAckTracker`. In this case you'd get a
+callback as soon as a message is delivered without preserving the original
+ordering.
+
+```python
+
+class MyFetcher(LogFetcher):
+
+    ...
+
+    def init(self, options):
+        self.ack_tracker = InstantAckTracker(ack_callback=self.message_acked)
+        return True
+
+    def message_acked(self, acked_message_bookmark):
+        pass
+```
+
+While ConsecutiveAckTracker() seems to provide a much more useful service,
+but InstantAckTracker() performs better, as it does not have to track
+acknowledgements of individual messages.
+
+The most complex scenario is implemented by BatchedAckTracker, this allows
+you to track the acks for individual messages, as they happen, not enforcing
+any kind of ordering.
+
+```python
+
+class MyFetcher(LogFetcher):
+
+    ...
+
+    def init(self, options):
+        self.ack_tracker = BatchedAckTracker(timeout=500, batch_size=100,
+                                             batched_ack_callback=self.messages_acked)
+        return True
+
+    def messages_acked(self, acked_message_bookmarks):
+        pass
+```
+
+BatchedAckTracker would call your callback every now and then, as specified
+by the `timeout` argument in milliseconds. `batch_size` specifies the number of
+outstanding messages at a time. With this interface it's quite easy to
+perform acknowledgements back to the source interface where per-message
+acknowledgements are needed (e.g. Google PubSub).
+
+### Creating a syslog-ng a LogSource based Source plugin
+
+While LogFetcher gives us a convinient interface for fetching messages from
+backend services via blocking APIs, it is limited to performing the fetching
+operation in a sequential manner: you fetch a batch of messages, feed the
+syslog-ng pipeline with those messages and then repeat.
+
+LogSource is more low-level but allows the use of an asynchronous framework
+(asyncio, etc) to perform message fetching along multiple threads of
+execution.
+
+This sample uses asyncio to generate two independent sequences of messages,
+one is generated every second, the other is every 1.5 seconds, running
+concurrently via an asyncio event loop.
```

Comment:
[optional]

LogFetcher, asyncio and LogSource are source codes, so they could be backticked.
```suggestion
While `LogFetcher` gives us a convinient interface for fetching messages from
backend services via blocking APIs, it is limited to performing the fetching
operation in a sequential manner: you fetch a batch of messages, feed the
syslog-ng pipeline with those messages and then repeat.

`LogSource` is more low-level but allows the use of an asynchronous framework
(`asyncio`, etc) to perform message fetching along multiple threads of
execution.

This sample uses `asyncio` to generate two independent sequences of messages,
one is generated every second, the other is every 1.5 seconds, running
concurrently via an `asyncio` event loop.
```